In [108]:
#manipulate dataframes in python
import pandas as pd

#make API calls with python
import requests

#allows us to store results of API call cleanly
import json

import math

In [47]:
# Sample variables
year='2018'
dsource='pep'
dname='charagegroups'
cols='GEONAME,POP,SEX,SUMLEV'
state='42'
county='017,029,045,091,101'

In [11]:
# Construct Based API
base_url = f'https://api.census.gov/data/{year}/{dsource}/{dname}'


In [14]:
#put your census API key here
api_key = "a45a33bca80b5e1907adc4587be5c346b897dda7"
#construct the API call we will use
data_url = f'{base_url}?get={cols}&for=county:{county}&in=state:{state}\
&key={api_key}'

In [16]:
#call the API and collect the response
response = requests.get(data_url)

In [17]:
print(response.text)

[["GEONAME","POP","SEX","SUMLEV","state","county"],
["Bucks County, Pennsylvania","628195","0","050","42","17"],
["Bucks County, Pennsylvania","308395","1","050","42","17"],
["Bucks County, Pennsylvania","319800","2","050","42","17"],
["Chester County, Pennsylvania","522046","0","050","42","29"],
["Chester County, Pennsylvania","257214","1","050","42","29"],
["Chester County, Pennsylvania","264832","2","050","42","29"],
["Delaware County, Pennsylvania","564751","0","050","42","45"],
["Delaware County, Pennsylvania","271816","1","050","42","45"],
["Delaware County, Pennsylvania","292935","2","050","42","45"],
["Montgomery County, Pennsylvania","828604","0","050","42","91"],
["Montgomery County, Pennsylvania","403268","1","050","42","91"],
["Montgomery County, Pennsylvania","425336","2","050","42","91"],
["Philadelphia County, Pennsylvania","1584138","0","050","42","101"],
["Philadelphia County, Pennsylvania","749572","1","050","42","101"],
["Philadelphia County, Pennsylvania","834566","

"item": {
      "0": "Both Sexes",
      "1": "Male",
      "2": "Female"

In [18]:
# Try with place code
state='36'
place='51000'

data_url2= f'{base_url}?get={cols}&for=place:{county}&in=state:{state}\
&key={api_key}'

In [19]:
response = requests.get(data_url2)

In [20]:
print(response.text)

error: unknown/unsupported geography heirarchy


## Place FIPS Code is not supported.

Then, I want to know how Census Quick Fact come up with the different Sex prediction
For example, NYC has 52.3% of Female population, and source is hoursing data set and ACA.

https://www.census.gov/quickfacts/newyorkcitynewyork

Population Estimates Program

Sex estimates of the population are produced for the United States, states, and counties, as well as for the Commonwealth of Puerto Rico and its municipios (county-equivalents for Puerto Rico). Sex is based on the biological attributes of men and women (chromosomes, anatomy, and hormones).

For more information, go to Age and Sex

American Community Survey

Sex estimates of the population are produced for places, zona urbanas and comunidades (place-equivalents for Puerto Rico), and minor civil divisions. The sex data collected on the forms are aggregated and provide the number of males and females in the population. These data are needed to interpret most social and economic characteristics used to plan and analyze programs and policies. Data about sex are critical because so many federal programs must differentiate between males and females.

For more information, go to ACS subject definitions "Sex"


Data users should be aware of methodology differences that may exist between different data sources.


## Therefore, I assume that the percentage of NYC female comes from NYC county population or State population.

Try with NY county 
From Wiki, Five of New York's counties are each coextensive with New York City's five boroughs and do not have county governments. They are New York County (Manhattan), Kings County (Brooklyn), Bronx County (The Bronx), Richmond County (Staten Island), and Queens County (Queens).

In [22]:
# LEt's check


state='36'
county='061,047,005,085,081'

In [26]:

data_url3= f'{base_url}?get={cols}&for=county:{county}&in=state:{state}\
&key={api_key}'

In [27]:
response = requests.get(data_url3)

In [28]:
print(response.text)

[["GEONAME","POP","SEX","SUMLEV","state","county"],
["Bronx County, New York","1432132","0","050","36","5"],
["Bronx County, New York","674900","1","050","36","5"],
["Bronx County, New York","757232","2","050","36","5"],
["Kings County, New York","2582830","0","050","36","47"],
["Kings County, New York","1223764","1","050","36","47"],
["Kings County, New York","1359066","2","050","36","47"],
["New York County, New York","1628701","0","050","36","61"],
["New York County, New York","771151","1","050","36","61"],
["New York County, New York","857550","2","050","36","61"],
["Queens County, New York","2278906","0","050","36","81"],
["Queens County, New York","1105389","1","050","36","81"],
["Queens County, New York","1173517","2","050","36","81"],
["Richmond County, New York","476179","0","050","36","85"],
["Richmond County, New York","230765","1","050","36","85"],
["Richmond County, New York","245414","2","050","36","85"]]


In [32]:
data=response.json()
df=pd.DataFrame(data[1:], columns=data[0])

In [33]:
df

GEONAME      POP SEX SUMLEV state county
0      Bronx County, New York  1432132   0    050    36      5
1      Bronx County, New York   674900   1    050    36      5
2      Bronx County, New York   757232   2    050    36      5
3      Kings County, New York  2582830   0    050    36     47
4      Kings County, New York  1223764   1    050    36     47
5      Kings County, New York  1359066   2    050    36     47
6   New York County, New York  1628701   0    050    36     61
7   New York County, New York   771151   1    050    36     61
8   New York County, New York   857550   2    050    36     61
9     Queens County, New York  2278906   0    050    36     81
10    Queens County, New York  1105389   1    050    36     81
11    Queens County, New York  1173517   2    050    36     81
12  Richmond County, New York   476179   0    050    36     85
13  Richmond County, New York   230765   1    050    36     85
14  Richmond County, New York   245414   2    050    36     85

In [37]:
df = df.astype(dtype={'POP':'int64'})

In [43]:
df[df['SEX']=='2']['POP'].sum()/df[df['SEX']=='0']['POP'].sum()

0.5230278370061824

In [44]:
df[df['SEX']=='0']['POP'].sum()

8398748

When I checked NYC CITY population from City dataset, population is also 8398748.

## Using ACS5

https://api.census.gov/data/2018/acs/acs5/subject?get=NAME,S0101_C01_001E&for=place:35536&in=state:01&key=YOUR_KEY_GOES_HERE

S0101_C01_001E	Estimate!!Total!!Total population <br>
S0101_C03_001E	Estimate!!Male!!Total population<br>
S0101_C05_001E	Estimate!!Female!!Total population<br>


In [164]:
year='2018'
dsource='acs'
dname='acs5'
cols='NAME,S0101_C01_001E,S0101_C03_001E,S0101_C05_001E'
state='36'
place='51000'

base_url4 = f'https://api.census.gov/data/{year}/{dsource}/{dname}/subject'

data_url4= f'{base_url4}?get={cols}&for=place:{place}&in=state:{state}\
&key={api_key}'

response4 = requests.get(data_url4)
print(response4.text)

[["NAME","S0101_C01_001E","S0101_C03_001E","S0101_C05_001E","state","place"],
["New York city, New York","8443713","4025097","4418616","36","51000"]]


## It is working for Gender

In [84]:
# for getting place ID, I will pull up wave1 data.
coltype1 = {'STATE':object,
           'PLACE':object}
wave1= pd.read_excel("wave1.xlsx", index_col=0,dtype=coltype1)

In [85]:
wave1.head()

STATE  PLACE  CITY_FIPS_CODE  Year             NAME   STNAME  \
0    01  00124          100124  2000   Abbeville city  Alabama   
1    01  00460          100460  2000  Adamsville city  Alabama   
2    01  00484          100484  2000     Addison town  Alabama   
3    01  00676          100676  2000       Akron town  Alabama   
4    01  00820          100820  2000   Alabaster city  Alabama   

   NYU_500_Largest  Pop_Estimate  
0                0          2985  
1                0          5021  
2                0           701  
3                0           485  
4                0         24116

In [86]:
# To get place Code, I need to divide by STATE because input is different.

wave1[wave1['STATE']=='01']['PLACE']

0         00124
1         00460
2         00484
3         00676
4         00820
          ...  
156416    83400
156417    83640
156418    83664
156419    83976
156420    84096
Name: PLACE, Length: 8749, dtype: object

In [88]:
# Using Set, we can find unique STATE FIPD CODE

state_fips = set(wave1['STATE'])

# It is working now for each city

#### AGE

S0101_C02_002E	Estimate!!Percent!!Total population!!AGE!!Under 5 years
S0101_C02_003E	Estimate!!Percent!!Total population!!AGE!!5 to 9 years
S0101_C02_004E	Estimate!!Percent!!Total population!!AGE!!10 to 14 years
S0101_C02_005E	Estimate!!Percent!!Total population!!AGE!!15 to 19 years
S0101_C02_006E	Estimate!!Percent!!Total population!!AGE!!20 to 24 years
S0101_C02_007E	Estimate!!Percent!!Total population!!AGE!!25 to 29 years
S0101_C02_008E	Estimate!!Percent!!Total population!!AGE!!30 to 34 years
S0101_C02_009E	Estimate!!Percent!!Total population!!AGE!!35 to 39 years
S0101_C02_010E	Estimate!!Percent!!Total population!!AGE!!40 to 44 years
S0101_C02_011E	Estimate!!Percent!!Total population!!AGE!!45 to 49 years
S0101_C02_012E	Estimate!!Percent!!Total population!!AGE!!50 to 54 years
S0101_C02_013E	Estimate!!Percent!!Total population!!AGE!!55 to 59 years
S0101_C02_014E	Estimate!!Percent!!Total population!!AGE!!60 to 64 years
S0101_C02_015E	Estimate!!Percent!!Total population!!AGE!!65 to 69 years
S0101_C02_016E	Estimate!!Percent!!Total population!!AGE!!70 to 74 years
S0101_C02_017E	Estimate!!Percent!!Total population!!AGE!!75 to 79 years
S0101_C02_018E	Estimate!!Percent!!Total population!!AGE!!80 to 84 years
S0101_C02_019E	Estimate!!Percent!!Total population!!AGE!!85 years and over

#### SEX
S0101_C04_001E	Estimate!!Percent Male!!Total population 
S0101_C06_001E	Estimate!!Percent Female!!Total population

--> these are for percentage of sub category. So, it comes out as -88888888, which is coded as none type. So, I will get the total population of each sex and get percentage.



In [172]:
year='2018'
dsource='acs'
dname='acs5'
cols='NAME,S0101_C01_001E,S0101_C02_002E,S0101_C02_003E,S0101_C02_004E,S0101_C02_005E,S0101_C02_006E,\
S0101_C02_007E,S0101_C02_008E,S0101_C02_009E,S0101_C02_010E,S0101_C02_011E,S0101_C02_012E,S0101_C02_013E,\
S0101_C02_014E,S0101_C02_015E,S0101_C02_016E,S0101_C02_017E,S0101_C02_018E,S0101_C02_019E,\
S0101_C03_001E,S0101_C05_001E'
state='36'
place='51000'

base_url4 = f'https://api.census.gov/data/{year}/{dsource}/{dname}/subject'

data_url4= f'{base_url4}?get={cols}&for=place:{place}&in=state:{state}\
&key={api_key}'

response4 = requests.get(data_url4)
print(response4.json()[1])

['New York city, New York', '8443713', '6.5', '5.6', '5.5', '5.4', '6.8', '9.3', '8.5', '7.2', '6.5', '6.4', '6.4', '6.1', '5.6', '4.6', '3.3', '2.5', '1.8', '1.9', '4025097', '4418616', '36', '51000']


In [175]:
len(wave1["PLACE"])

370615

In [ ]:
master=[]
year='2018'
dsource='acs'
dname='acs5'
cols=cols='NAME,S0101_C01_001E,S0101_C02_002E,S0101_C02_003E,S0101_C02_004E,S0101_C02_005E,S0101_C02_006E,\
S0101_C02_007E,S0101_C02_008E,S0101_C02_009E,S0101_C02_010E,S0101_C02_011E,S0101_C02_012E,S0101_C02_013E,\
S0101_C02_014E,S0101_C02_015E,S0101_C02_016E,S0101_C02_017E,S0101_C02_018E,S0101_C02_019E,\
S0101_C03_001E,S0101_C05_001E'
state_fips = set(wave1['STATE'])
for state in state_fips:
    place_fips = wave1[wave1['STATE']==state]['PLACE']
    print(state)
    for place in place_fips:
        print(place)
        base_url5 = f'https://api.census.gov/data/{year}/{dsource}/{dname}/subject'
        #print(base_url5)
        data_url5= f'{base_url5}?get={cols}&for=place:{place}&in=state:{state}&key={api_key}'
        #print(data_url5)
        response5 = requests.get(data_url5)
        data=response5.json()[1]
        master.append(data)

35
01780
02000
03400
05220
05780
06270
06480
06970
07880
08580
11800
12150
12500
12850
13970
15440
15720
16280
16420
17050
17680
17960
19150
20270
20480
20620
21180
22020
22380
22720
23000
24610
25170
25380
25450
25800
26570
26780
27340
28460
30210
30490
30770
31330
31820
32520
33290
33710
33850
35040
35320
37840
39380
39940
42040
42180
43370
43930
44420
44490
46310
47080
47570
48060
48620
50160
50300
50370
55620
56180
59260
60870
62060
62200
62620
63460
64930
65070
65210
65280
68150
70270
70500
70670
71020
73260
73540
74800
75640
76200
76480
76620
77250
77880
79840
79910
79980
82570
83200
83340
85160
85300
01780
02000
03400
05220
05780
06270
06480
06970
07880
08580
11800
12150
12500
12850
13970
15440
15720
16280
16420
17050
17680
17960
19150
20270
20480
20620
21180
22020
22380
22720
23000
24610
25170
25380
25450
25800
26570
26780
27340
28460
30210
30490
30770
31330
31820
32520
33290
33710
33850
35040
35320
37840
39380
39940
42040
42180
43370
43930
44420
44490
46310
47080
47570
48060
4